### PR TITLE
console and etw should work together in lasterror cases

### PR DESCRIPTION
**Pull request by anporumb**

### CODE REVIEW

File: src/etwlogger_driver.c [hunk 1]
```diff
@@ -113,26 +113,98 @@ static char* lastErrorToString(DWORD lastError)
...
+    if (event_write_result != ERROR_SUCCESS)
+    {
+        if (InterlockedCompareExchange(&stopPrintfing, 1, 0) == 0)
+        {
+            (void)printf("failed in EventWriteLogErrorEvent: %lu. Further failing calls to EventWriteLog... will not result in printf\r\n", event_write_result);
+        }
+    }
+    else
+    {
+        InterlockedExchange(&stopPrintfing, 0);
```
Comment (dcristoloveanu) [inline src/etwlogger_driver.c]: >InterlockedExchange [](start = 8, length = 19)

(void)

File: src/etwlogger_driver.c [hunk 2]
```diff
@@ -113,26 +113,98 @@ static char* lastErrorToString(DWORD lastError)
...
+    if (event_write_result != ERROR_SUCCESS)
+    {
+        if (InterlockedCompareExchange(&stopPrintfing, 1, 0) == 0)
+        {
+            (void)printf("failed in EventWriteLogLastError: %lu. Further failing calls to EventWriteLog... will not result in printf\r\n", event_write_result);
+        }
+    }
+    else
+    {
+        InterlockedExchange(&stopPrintfing, 0);
```
Comment (dcristoloveanu) [inline src/etwlogger_driver.c]: >InterlockedExchange [](start = 8, length = 19)

(void)

File: src/etwlogger_driver.c [hunk 3]
```diff
@@ -260,46 +285,18 @@ void etwlogger_log(LOG_CATEGORY log_category, const char* file, const char* func
...
-                /*fallback on printf...*/
-                if (InterlockedCompareExchange(&stopPrintfing, 1, 0) == 0)
-                {
-                    (void)printf("failed in EventWriteLogErrorEvent: %lu. Further failing calls to EventWriteLogErrorEvent will not result in printf\r\n", event_write_result);
-                }
-            }
-            else
-            {
-                (void)InterlockedExchange(&stopPrintfing, 0);
-            }
```
Comment (dcristoloveanu) [inline src/etwlogger_driver.c]: Love this extraction

File: src/etwlogger_driver.c [hunk 4]
```diff
@@ -113,26 +113,98 @@ static char* lastErrorToString(DWORD lastError)
...
+    {
+        /*do nothing, already registered/attempted*/
+    }
+}
+
+/*the below interlocked variable initially set to 0*/
+/* At the first error returned by EventWriteLogInfoEvent or by EventWriteLogErrorEvent a printf will be performed and the variable will be switched to 1. */
+/* while the variable is set to "1" no fallback printf will be performed */
+/* at the first non-erroneous EventWriteLogInfoEvent or EventWriteLogErrorEvent the printf fallback will be re-enabled */
+static volatile LONG stopPrintfing = 0;
```
Comment (dcristoloveanu) [inline src/etwlogger_driver.c]: >stopPrintfing [](start = 21, length = 13)

/poetry This is just a matter of preference. I would normally have liked the variable name to represent what it enables (printfing_enabled) not the action (stop). But this is so small that maybe I should not even made the comment :-).

File: src/etwlogger_driver.c (hunk 1 above)
Comment (anporumb) [reply src/etwlogger_driver.c]: added to all InterlockedExchanges

---
In reply to: [319620175](https://github.com/Azure/azure-c-shared-utility/pull/367#discussion_r319620175) [](ancestors = 319620175)

File: src/etwlogger_driver.c (hunk 4 above)
Comment (anporumb) [reply src/etwlogger_driver.c]: renamed to "doPrintf" and changed the logic to positive

---
In reply to: [319621136](https://github.com/Azure/azure-c-shared-utility/pull/367#discussion_r319621136) [](ancestors = 319621136)

**Pull request merged**